awx/main/models/inventory.py

```diff
@@ -820,9 +820,47 @@ def _get_related_jobs(self):
 
 
 class HostMetric(models.Model):
-    hostname = models.CharField(primary_key=True, max_length=512)
+    hostname = models.CharField(unique=True, max_length=512)
     first_automation = models.DateTimeField(auto_now_add=True, null=False, db_index=True, help_text=_('When the host was first automated against'))
     last_automation = models.DateTimeField(db_index=True, help_text=_('When the host was last automated against'))
+    last_deleted = models.DateTimeField(null=True, db_index=True, help_text=_('When the host was last deleted'))
+    automated_counter = models.BigIntegerField(default=0, help_text=_('How many times was the host automated'))
+    deleted_counter = models.BigIntegerField(default=0, help_text=_('How many times was the host deleted'))
+    deleted = models.BooleanField(
+        default=False, help_text=_('Boolean flag saying whether the host is deleted and therefore not counted into the subscription consumption')
```

Comment:
All deletion related fields, `last_deleted`, `deleted_counter`, and `deleted` are going to be very challenging to satisfy from a performance standpoint. We have a task `delete_inventory` that is the asynchronous task for deleting an inventory and its contents. It breaks frequently, and this will add additional logic to it. I can start talking out some things we need to do for that...

We should absolutely make HostMetric updates _outside of_ the database transaction where we delete the inventory objects, because I would prefer to move along with partially incorrect data as opposed to revert the inventory deletion which would be a bug that significantly interferes with the user experience.

I'm a little more concerned with the implications for inventory imports. The entire beautify of `HostMetric` as it exists now is that we only had to update it _reactively_ as jobs finished. With the addition of these fields we will have to _proactively_ update the data as a side-effect of the inventory update logic to save the content.

What if a previously-automated `Host` is _renamed_? Should the `HostMetric` model reflect that the host was deleted? If yes, then this is a costly decision.

Reply:
cc @Ladas

Reply:
@AlanCoding right now there is no cascade delete or any connection to Inventory.

Soft-Deletes from HostMetric will be done manually through UI/API and automatically via the scheduled task. (so e.g. automatically marking HostMetric rows as soft-deleted after a year with no automation, giving user the ability to tweak the timeframe via settings)

Then if the host is automated again, the deleted flag is reverted. So adding/removing from inventory will not have any effect here now.

------

Connection to Inventory is a future TBD, so e.g. we might attempt to delete the host if it will be deleted from inventory and it's a last occurence of the host. But there is no such requirement now. (the only connection here is an periodic job, that will compute how many inventories is the HostMetric row in, so customers can order by this in the UI. Later we could leverage it for scheduled delete too...)

Reply:
@AlanCoding 
I tested the bulk_create:
In Django 3.2 (used in awx), the bulk_create returns all objects including ignored. And even ignored have updated values. So it's not possible to recognize what was and what wasn't updated without additional query => excluding from update is not worth. 
Regarding [django 4.2 doc](https://docs.djangoproject.com/en/4.2/ref/models/querysets/#bulk-create) the `bulk_create` offers `update_fields` and `update_conflicts` (not clear about effectivity). So in order to keep usage of ORM and be more effective Django update is required

Reply:
@AlanCoding in current work, renaming is not directly reflected in HostMetrics. But customer can (soft-)delete HostMetric manually (through API/UI) or there will be cleanup scheduled task (in next PR) which will soft-delete records after certain time with no updates